ambry-tools/src/test/java/com/github/ambry/clustermap/HelixBootstrapUpgradeToolTest.java

```diff
@@ -0,0 +1,399 @@
+  }
+
+  /**
+   * Write the layout files out from the constructed in-memory hardware and partition layouts; use the bootstrap tool
+   * to update the contents in Helix; verify that the information is consistent between the two.
+   * @param expectedResourceCount number of resources expected in Helix for this cluster in each datacenter.
+   * @throws IOException if a file read error is encountered.
+   * @throws JSONException if a JSON parse error is encountered.
+   */
+  private void writeBootstrapOrUpgradeAndVerify(long expectedResourceCount) throws IOException, JSONException {
+    Utils.writeJsonToFile(zkJson, zkLayoutPath);
+    Utils.writeJsonToFile(testHardwareLayout.getHardwareLayout().toJSONObject(), hardwareLayoutPath);
+    Utils.writeJsonToFile(testPartitionLayout.getPartitionLayout().toJSONObject(), partitionLayoutPath);
+    HelixBootstrapUpgradeTool.bootstrapOrUpgrade(hardwareLayoutPath, partitionLayoutPath, zkLayoutPath, "DC1");
+    verifyEquivalencyWithStatic(testHardwareLayout.getHardwareLayout(), testPartitionLayout.getPartitionLayout(),
+        expectedResourceCount);
+  }
+
+  /**
+   * Verify that the information in Helix and the information in the static clustermap are equivalent.
+   * @param hardwareLayout the {@link HardwareLayout} of the static clustermap.
+   * @param partitionLayout the {@link PartitionLayout} of the static clustermap.
+   * @param expectedResourceCount the expected number of resources in Helix.
+   */
+  private void verifyEquivalencyWithStatic(HardwareLayout hardwareLayout, PartitionLayout partitionLayout,
+      long expectedResourceCount) {
+    String clusterName = hardwareLayout.getClusterName();
+    for (Datacenter dc : hardwareLayout.getDatacenters()) {
+      ZkInfo zkInfo = dcsToZkInfo.get(dc.getName());
+      Assert.assertNotNull(zkInfo);
+      ZKHelixAdmin admin = new ZKHelixAdmin("localhost:" + zkInfo.port);
+      Assert.assertTrue(admin.getClusters().contains(clusterName));
+      verifyResourcesAndPartitionEquivalencyInDc(dc, clusterName, partitionLayout, expectedResourceCount);
+      verifyDataNodeAndDiskEquivalencyInDc(dc, clusterName, hardwareLayout, partitionLayout);
+    }
+  }
+
+  /**
+   * Verify that the hardware layout information is in sync - which includes the node and disk information. Also verify
+   * that the replicas belonging to disks are in sync between the static cluster map and Helix.
+   * @param dc the datacenter whose information is to be verified.
+   * @param clusterName the cluster to be verified.
+   * @param hardwareLayout the {@link HardwareLayout} of the static clustermap.
+   * @param partitionLayout the {@link PartitionLayout} of the static clustermap.
+   */
+  private void verifyDataNodeAndDiskEquivalencyInDc(Datacenter dc, String clusterName, HardwareLayout hardwareLayout,
+      PartitionLayout partitionLayout) {
+    ClusterMapManager staticClusterMap = new ClusterMapManager(partitionLayout);
```

Comment:
its ironic that hardwareLayout is never used to check for data node and disk equivalency ;)